File: University/Tasks/Solution/Labs/Sharp/ZoologicGarden/Animal.cs
using System;

namespace Labs.Sharp.ZoologicGarden
{
    public abstract class Animal : IAnimalA, IAnimalB
    {
        public string Name { get; private set; }
        public double Weight { get; private set; }
        public Color Color { get; private set; }
        public DangerType Danger { get; private set; }
        public AnimalType Type { get; private set; }

        public abstract string Say();

        protected abstract double LengthStep { get; }

        public Animal(AnimalType type, string name, double weight, Color color, DangerType danger)
        {
            Name = name;
            Weight = weight;
            Color = color;
            Type = type;
            Danger = danger;
        }

        public string MakeStep(int count)
        {
            var length = count * LengthStep;

            return $"{Name} сделал {count} шаг(ов). При этом прошёл {length} метров.";
        }
        public override string ToString()
        {
            return $"Type: {Type}; Name: {Name}; Weight: {Weight}; Color: {Color}; Length of step: {LengthStep}, Type of danger {Danger}.";
        }
        public string Say(string @string)
        {
            return $"{Name}: {@string}";
        }

        public string Hello1()
        {
            return "AB: Hello1";
        }

        public string Hello2()
        {
            var a = ((IAnimalA)this).Hello2();
            var b = ((IAnimalB)this).Hello2();
            return a + b;
        }
        string IAnimalA.Hello2()
        {
            return "A: Hello2";
        }
        string IAnimalB.Hello2()
        {
            return "B: Hello2";
        }
    }
}

File: University/Tasks/Solution/Labs/Plus/ZoologicGarden/Animal.cpp
#include "Animal.h"

#pragma region Public
Animal::Animal(char* name, double weight, Color color, AnimalType type)
{
	int length = std::strlen(name);
	_name = new char[length];

	for (int i = 0; i < length; i++)
	{
		_name[i] = name[i];
	}

	// Конец строки.
	_name[length] = '\0';

	_weight = weight;
	_color = color;
	_type = type;
}
Animal::~Animal()
{
	delete[] _name;
}

void Animal::MakeStep(int count)
{
	int length = GetLengthStep() * count;

	std::cout << _name << " make " << count << " step(s). Metrs: " << length << "." << std::endl;
}
void Animal::Print()
{
	std::cout << "Type: " << TypeToString[_type] << "; Name: " << _name << "; Weight = " << _weight <<
		"; Color: " << ColorToString[_color] << "; Length of step: " << GetLengthStep() << std::endl;
}
#pragma endregion

#pragma region Private
void Animal::SayWord(char* str)
{
	std::cout << _name << ": " << str << std::endl;
}
#pragma endregion